src/dragonball/src/dbs_pci/src/root_device.rs

```diff
@@ -0,0 +1,411 @@
+            1 => {
+                // Each PCI bus consumes 1MB of MMIO address range.
+                assert!(mmios[0].1 >= u64::from(root.max_bus_id) << 20);
+                root.mmio_base = mmios[0].0;
+                root.mmio_size = mmios[0].1;
+                found = true;
+            }
+            _ => return Err(Error::NoResources),
+        }
+
+        // At lease one of IO port or MMIO must be enabled.
+        if !found {
+            return Err(Error::NoResources);
+        }
+
+        Ok(root)
+    }
+
+    /// Activate the PCI root device, getting ready to handle PCI configuration space accesses from
+    /// the guest.
+    pub fn activate<I: IoManagerContext>(root: Arc<PciRootDevice>, io_ctx: &mut I) -> Result<()> {
+        let mut tx = io_ctx.begin_tx();
+        if let Err(e) = io_ctx.register_device_io(&mut tx, root.clone(), &root.resources) {
+            io_ctx.cancel_tx(tx);
+            Err(Error::ActivateFailure(e))
+        } else {
+            io_ctx.commit_tx(tx);
+            Ok(())
+        }
+    }
+
+    /// Add a PCI bus instance to be managed by the root device.
+    pub fn add_bus(&self, bus: Arc<PciBus>, id: u8) -> Result<()> {
+        if id > self.max_bus_id {
+            return Err(Error::InvalidBusId(id));
+        }
+        // Don't expect poisoned lock here.
+        self.state
+            .write()
+            .expect("poisoned lock for PCI root device")
+            .buses
+            .insert(id as u32, bus);
+
+        Ok(())
+    }
+
+    /// Get a PCI bus instance by bus id.
+    pub fn get_bus_by_id(&self, id: u8) -> Option<Arc<PciBus>> {
+        if id > self.max_bus_id {
+            return None;
+        }
+
+        // Don't expect poisoned lock here.
+        self.state
+            .read()
+            .expect("poisoned lock for PCI root device")
+            .buses
+            .get(&(id as u32))
+            .cloned()
+    }
+
+    /// Get PCI root device resources
+    #[cfg(target_arch = "aarch64")]
+    pub fn get_device_resources(&self) -> DeviceResources {
+        self.resources.clone()
+    }
+}
+
+impl DeviceIo for PciRootDevice {
+    /// At present, only arm will go to this process, because the x86 pci root device does
+    /// not allocate mmio resources. This process does not involve the arm proprietary
+    /// interface, so it does not use the aarch64 macro to wrap.
+    fn read(&self, base: IoAddress, offset: IoAddress, data: &mut [u8]) {
+        let offset = offset.raw_value();
+        let len = data.len();
+
+        // Only allow naturally aligned Dword, Word and Byte access.
+        if (len == 4 || len == 2 || len == 1) && offset as usize & (len - 1) == 0 {
+            // Do not expect poisoned lock here.
+            let state = self.state.read().unwrap();
+            let (b, d, f, o) = parse_mmio_address(offset);
+            if let Some(bus) = state.buses.get(&b) {
+                return bus.read_config(d, f, o | ((offset as u32) & 0x3), data);
```

Comment:
Could you set a const for `0x3` to describe what it is? Or leave some comments here.

Reply:
I found out that 0x3 is useless, so I remove them from the code.